bindgen/src/formatter.ts

```diff
@@ -22,51 +22,49 @@ import cp from "child_process";
 import { extend } from "./debug";
 const debug = extend("format");
 
-const FORMATTERS: Record<string, string[]> = {
-  eslint: ["npx", "eslint", "--fix", "--format=stylish"],
-  "clang-format": ["npx", "clang-format", "-i"],
-  // "typescript-checker": ["npx", "tsc", "--noEmit"],
-};
+export type Formatter = (cwd: string, filePaths: string[]) => void;
 
-export function addFormatter(name: string, commandAndArgs: string[]) {
-  if (name in FORMATTERS) {
-    throw new Error(`Cannot add ${name} as it's already there`);
-  } else {
-    FORMATTERS[name] = commandAndArgs;
+export function executeCommand(cwd: string, command: string, ...args: string[]) {
+  console.log(chalk.dim(command, ...args));
+  const result = cp.spawnSync(command, args, {
+    cwd,
+    encoding: "utf8",
+    stdio: "inherit",
+    shell: true,
+  });
+  if (result.error) throw result.error;
+  if (result.status) {
+    throw new Error(`Exited with status ${result.status}`)
   }
 }
 
-export function getFormatterNames() {
-  return Object.keys(FORMATTERS);
+export function createCommandFormatter(formatterName: string, [command, ...args]: string[]): Formatter {
+  const formatter: Formatter = (cwd, filePaths) => {
+    executeCommand(cwd, command, ...args, ...filePaths);
+  };
+  Object.defineProperty(formatter, "name", { value: formatterName });
+  return formatter;
 }
 
-export type FormatterName = string;
+export const clangFormatter = createCommandFormatter("clang", ["npx", "clang-format", "-i"]);
 
 export class FormatError extends Error {
-  constructor(formatterName: string, filePaths: string[]) {
-    super(`Failure when running the '${formatterName}' formatter on ${JSON.stringify(filePaths)}`);
+  constructor(formatterName: string, filePaths: string[], cause: Error) {
+    super(`Failure when running the '${formatterName}' formatter on ${JSON.stringify(filePaths)}: ${cause.message}`);
   }
 }
 
-export function format(formatterName: FormatterName, cwd: string, filePaths: string[]): void {
+export function format(formatter: Formatter, cwd: string, filePaths: string[]): void {
   if (filePaths.length === 0) {
-    debug(chalk.dim("Skipped running formatter '%s' (no files need it)"), formatterName);
+    debug(chalk.dim("Skipped running formatter '%s' (no files need it)"), formatter.name);
```

Comment:
Why isn't TS complaining about the use of `formatter.name` when the `Formatter` type doesn't have a `name` property? Won't this break if somebody passes a lambda to `format()`?

Ditto other usages of `.name`

(of course if you go back to commandAndArg lists, this won't matter any more)

Reply:
I believe every function has a `name` - it's just the empty string for an arrow function.

Reply:
D'oh! I forgot about that. Makes sense now.

Reply:
I'll add an assert for that.

Reply:
I wanted to ensure that a name is provided so printing it as part of running it won't just be an empty string.